docs/design-proposals/2022-01 Metrics-generator.md

```diff
@@ -0,0 +1,298 @@
+
+A couple of notable differences between the Tempo metrics-generator and the Cortex/Loki ruler:
+
+- The metrics-generator has to consume the ingress stream. Because of this, the metrics-generator can only generate metrics about data that is being ingested. I.e. it's not possible to generate metrics from previously ingested data or to backfill metrics.
+
+- The metrics-generator uses fixed processors, these are less flexible than a rule which can contain a user-defined custom query. On the other hand, these processors can perform calculations which can't be expressed in a query language. The processing done by the service graph processor for instance will be difficult to express in a query.
+
+## Components
+
+This section takes a more detailed look at the components involved in the path between ingesting traces and writing metrics.
+
+### Distributor
+
+The distributor is the entrypoint for Tempo writes: it will receives batches of spans and forwards them to the ingesters (using replication if enabled). With the metrics-generator in the system, the distributor will now also have to write data to the metrics-generator. The distributor will first write data to ingesters and if this was successful it will push the same data to the metrics-generator.
+
+Writing to the metrics-generator is on a _best effort basis_: even if writing to the metrics-generator fails the Tempo write is still considered successful. This is a trade-off to keep request handling simple: if writing to the ingester succeeds but writing to the metrics-generator fails, the distributor should also revert the ingester write. The logic to discard already ingested data is deemed too complex.
+
+This tradeoff will result in missing or incomplete metrics whenever the metrics-generator is not able to ingest some data. As the client will not be aware of this, it will not resend the request. Failed writes should be reported with a metric on the distributor which can alert an operator (e.g. `distributor_metrics_generator_pushes_failures_total`).
+
+#### Metrics-generator ring
+
+The distributor has to find metrics-generator instances present in the cluster. When multiple instances of the metrics-generator are running, the distributor should load balance writes across these instances. Load balancing has to be trace-aware: all spans with the same trace ID should consistently be sent to the same instance.
+
+To achieve this we propose using the [dskit ring](https://github.com/grafana/dskit/tree/main/ring) backed by memberlist. This will be the same mechanism as used by the ingesters. The distributor will shard requests across metrics-generator instances based upon the tokens they own.
```

Comment:
by default tempo uses memberlist but technically this should be configurable like all our other rings. i.e. if an operator wants to use consul or etcd to store this ring then it should be possible.

but overall +1 the choice to use the same ring mechanism we use for sharding writes to the ingesters.

Reply:
Good point, I don't see why this wouldn't work with any other supported KV store. I mentioned memberlist because it's the default in Tempo. I'll clarify this.